File: scripts/progression/playerProgression.js
// Global Player Progression System
import { storage } from '../storage-adapter.js';

export class PlayerProgression {
    constructor() {
        this.level = 1;
        this.xp = 0;
        this.totalXP = 0;
        this.rank = 'Novice';
        this.stats = {
            gamesPlayed: 0,
            totalScore: 0,
            totalLines: 0,
            totalTime: 0,
            favoriteMode: null,
            achievements: [],
            trophies: [],
            puzzlesCompleted: 0,
            dailyChallengesCompleted: 0,
            highestCombo: 0,
            totalTSpins: 0,
            totalTetris: 0,
            perfectClears: 0
        };
        this.unlocks = {
            themes: ['default'],
            music: ['classic'],
            backgrounds: ['space'],
            pieceStyles: ['neon'],
            effects: ['basic']
        };
        this.badges = [];
        this.loadProgress();
    }
    
    async loadProgress() {
        const saved = await storage.load('player_progression');
        if (saved) {
            Object.assign(this, saved);
        }
    }
    
    async saveProgress() {
        await storage.save('player_progression', {
            level: this.level,
            xp: this.xp,
            totalXP: this.totalXP,
            rank: this.rank,
            stats: this.stats,
            unlocks: this.unlocks,
            badges: this.badges
        });
    }
    
    addXP(amount, source = 'gameplay') {
        this.xp += amount;
        this.totalXP += amount;
        
        // Check for level up
        const requiredXP = this.getRequiredXP();
        while (this.xp >= requiredXP) {
            this.levelUp();
        }
        
        // Update rank
        this.updateRank();
        
        // Save progress
        this.saveProgress();
        
        return {
            xpGained: amount,
            currentXP: this.xp,
            level: this.level,
            rank: this.rank,
            source: source
        };
    }
    
    getRequiredXP() {
        // XP required for next level (exponential curve)
        return Math.floor(100 * Math.pow(1.5, this.level - 1));
    }
    
    levelUp() {
        const requiredXP = this.getRequiredXP();
        this.xp -= requiredXP;
        this.level++;
        
        // Check for unlocks
        const newUnlocks = this.checkUnlocks();
        
        // Trigger level up event
        this.onLevelUp(newUnlocks);
        
        return {
            newLevel: this.level,
            unlocks: newUnlocks
        };
    }
    
    updateRank() {
        const ranks = [
            { level: 1, name: 'Novice' },
            { level: 5, name: 'Apprentice' },
            { level: 10, name: 'Adept' },
            { level: 20, name: 'Expert' },
            { level: 30, name: 'Master' },
            { level: 40, name: 'Grandmaster' },
            { level: 50, name: 'Champion' },
            { level: 60, name: 'Legend' },
            { level: 75, name: 'Mythic' },
            { level: 100, name: 'Eternal' }
        ];
        
        for (let i = ranks.length - 1; i >= 0; i--) {
            if (this.level >= ranks[i].level) {
                this.rank = ranks[i].name;
                break;
            }
        }
    }
    
    checkUnlocks() {
        const unlocks = [];
        
        // Theme unlocks
        const themeUnlocks = [
            { level: 5, theme: 'cyberpunk', name: 'Cyberpunk Theme' },
            { level: 10, theme: 'retro', name: 'Retro Theme' },
            { level: 15, theme: 'nature', name: 'Nature Theme' },
            { level: 20, theme: 'minimal', name: 'Minimal Theme' },
            { level: 30, theme: 'galaxy', name: 'Galaxy Theme' },
            { level: 40, theme: 'matrix', name: 'Matrix Theme' },
            { level: 50, theme: 'rainbow', name: 'Rainbow Theme' }
        ];
        
        themeUnlocks.forEach(unlock => {
            if (this.level === unlock.level && !this.unlocks.themes.includes(unlock.theme)) {
                this.unlocks.themes.push(unlock.theme);
                unlocks.push({ type: 'theme', id: unlock.theme, name: unlock.name });
            }
        });
        
        // Music unlocks
        const musicUnlocks = [
            { level: 3, track: 'chiptune', name: 'Chiptune Music' },
            { level: 8, track: 'synthwave', name: 'Synthwave Music' },
            { level: 12, track: 'orchestral', name: 'Orchestral Music' },
            { level: 18, track: 'jazz', name: 'Jazz Remix' },
            { level: 25, track: 'metal', name: 'Metal Version' },
            { level: 35, track: 'lofi', name: 'Lo-Fi Beats' }
        ];
        
        musicUnlocks.forEach(unlock => {
            if (this.level === unlock.level && !this.unlocks.music.includes(unlock.track)) {
                this.unlocks.music.push(unlock.track);
                unlocks.push({ type: 'music', id: unlock.track, name: unlock.name });
            }
        });
        
        // Piece style unlocks
        const pieceUnlocks = [
            { level: 7, style: 'glass', name: 'Glass Pieces' },
            { level: 14, style: 'pixel', name: 'Pixel Art Pieces' },
            { level: 22, style: 'hologram', name: 'Hologram Pieces' },
            { level: 28, style: 'crystal', name: 'Crystal Pieces' },
            { level: 45, style: 'animated', name: 'Animated Pieces' }
        ];
        
        pieceUnlocks.forEach(unlock => {
            if (this.level === unlock.level && !this.unlocks.pieceStyles.includes(unlock.style)) {
                this.unlocks.pieceStyles.push(unlock.style);
                unlocks.push({ type: 'pieceStyle', id: unlock.style, name: unlock.name });
            }
        });
        
        // Effect unlocks
        const effectUnlocks = [
            { level: 6, effect: 'particles', name: 'Particle Effects' },
            { level: 11, effect: 'trails', name: 'Piece Trails' },
            { level: 16, effect: 'explosions', name: 'Line Clear Explosions' },
            { level: 24, effect: 'lightning', name: 'Lightning Effects' },
            { level: 32, effect: 'shatter', name: 'Shatter Effects' },
            { level: 55, effect: 'quantum', name: 'Quantum Effects' }
        ];
        
        effectUnlocks.forEach(unlock => {
            if (this.level === unlock.level && !this.unlocks.effects.includes(unlock.effect)) {
                this.unlocks.effects.push(unlock.effect);
                unlocks.push({ type: 'effect', id: unlock.effect, name: unlock.name });
            }
        });
        
        return unlocks;
    }
    
    onLevelUp(unlocks) {
        // Create level up notification
        const notification = {
            type: 'levelUp',
            level: this.level,
            rank: this.rank,
            unlocks: unlocks,
            timestamp: Date.now()
        };
        
        // Dispatch event
        window.dispatchEvent(new CustomEvent('playerLevelUp', { detail: notification }));
        
        // Show UI notification
        if (window.game && window.game.uiManager) {
            window.game.uiManager.showLevelUp(this.level, this.rank, unlocks);
        }
    }
    
    updateStats(gameResults) {
        this.stats.gamesPlayed++;
        this.stats.totalScore += gameResults.score || 0;
        this.stats.totalLines += gameResults.lines || 0;
        this.stats.totalTime += gameResults.time || 0;
        
        if (gameResults.combo > this.stats.highestCombo) {
            this.stats.highestCombo = gameResults.combo;
        }
        
        this.stats.totalTSpins += gameResults.tspins || 0;
        this.stats.totalTetris += gameResults.tetris || 0;
        
        if (gameResults.perfectClear) {
            this.stats.perfectClears++;
        }
        
        // Update favorite mode
        if (gameResults.mode) {
            if (!this.stats.modeCount) {
                this.stats.modeCount = {};
            }
            this.stats.modeCount[gameResults.mode] = (this.stats.modeCount[gameResults.mode] || 0) + 1;
            
            // Find most played mode
            let maxCount = 0;
            let favoriteMode = null;
            for (const [mode, count] of Object.entries(this.stats.modeCount)) {
                if (count > maxCount) {
                    maxCount = count;
                    favoriteMode = mode;
                }
            }
            this.stats.favoriteMode = favoriteMode;
        }
        
        this.saveProgress();
    }
    
    calculateGameXP(gameResults) {
        let xp = 0;
        
        // Base XP from score
        xp += Math.floor(gameResults.score / 100);
        
        // Lines cleared
        xp += gameResults.lines * 10;
        
        // Special moves
        xp += gameResults.tspins * 50;
        xp += gameResults.tetris * 30;
        xp += gameResults.perfectClears * 100;
        
        // Combo bonus
        if (gameResults.combo >= 10) xp += 100;
        if (gameResults.combo >= 20) xp += 200;
        
        // Mode multipliers
        const modeMultipliers = {
            classic: 1.0,
            sprint: 1.2,
            marathon: 1.5,
            puzzle: 1.3,
            battle: 1.4,
            powerup: 1.1,
            zen: 0.8
        };
        
        const multiplier = modeMultipliers[gameResults.mode] || 1.0;
        xp = Math.floor(xp * multiplier);
        
        // First win of the day bonus
        if (this.isFirstWinOfDay()) {
            xp *= 2;
        }
        
        return xp;
    }
    
    isFirstWinOfDay() {
        const today = new Date().toISOString().split('T')[0];
        const lastWin = this.stats.lastWinDate;
        
        if (lastWin !== today) {
            this.stats.lastWinDate = today;
            return true;
        }
        return false;
    }
    
    addBadge(badge) {
        if (!this.badges.find(b => b.id === badge.id)) {
            this.badges.push({
                ...badge,
                unlockedAt: Date.now()
            });
            this.saveProgress();
            return true;
        }
        return false;
    }
    
    getProgressToNextLevel() {
        const required = this.getRequiredXP();
        return {
            current: this.xp,
            required: required,
            percentage: Math.floor((this.xp / required) * 100)
        };
    }
    
    getPlayerCard() {
        return {
            name: this.getPlayerName(),
            level: this.level,
            rank: this.rank,
            totalXP: this.totalXP,
            badges: this.badges.length,
            stats: {
                games: this.stats.gamesPlayed,
                avgScore: Math.floor(this.stats.totalScore / Math.max(1, this.stats.gamesPlayed)),
                favoriteMode: this.stats.favoriteMode,
                achievements: this.stats.achievements.length
            },
            unlocks: {
                themes: this.unlocks.themes.length,
                music: this.unlocks.music.length,
                effects: this.unlocks.effects.length
            }
        };
    }
    
    getPlayerName() {
        // Get from settings or use default
        return localStorage.getItem('tetris_player_name') || 'Player';
    }
    
    reset() {
        this.level = 1;
        this.xp = 0;
        this.totalXP = 0;
        this.rank = 'Novice';
        this.stats = {
            gamesPlayed: 0,
            totalScore: 0,
            totalLines: 0,
            totalTime: 0,
            favoriteMode: null,
            achievements: [],
            trophies: [],
            puzzlesCompleted: 0,
            dailyChallengesCompleted: 0,
            highestCombo: 0,
            totalTSpins: 0,
            totalTetris: 0,
            perfectClears: 0
        };
        this.unlocks = {
            themes: ['default'],
            music: ['classic'],
            backgrounds: ['space'],
            pieceStyles: ['neon'],
            effects: ['basic']
        };
        this.badges = [];
        this.saveProgress();
    }
}

export const playerProgression = new PlayerProgression();

File: android-app/app/src/main/java/com/tetris/modern/rl/progression/PlayerProgression.kt
package com.tetris.modern.rl.progression

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.pow

@Singleton
class PlayerProgression @Inject constructor() {
    
    companion object {
        const val MAX_LEVEL = 100
        const val XP_PER_LEVEL_BASE = 1000
        const val XP_PER_LEVEL_MULTIPLIER = 1.15f
        
        // XP Rewards
        const val XP_PER_LINE = 10
        const val XP_PER_TETRIS = 100
        const val XP_PER_T_SPIN = 50
        const val XP_PER_PERFECT_CLEAR = 200
        const val XP_PER_COMBO = 15
        const val XP_PER_LEVEL_UP = 100
        const val XP_DAILY_CHALLENGE = 500
        const val XP_ACHIEVEMENT = 250
    }
    
    data class ProgressionState(
        val level: Int = 1,
        val currentXP: Int = 0,
        val totalXP: Int = 0,
        val xpToNextLevel: Int = XP_PER_LEVEL_BASE,
        val rank: Rank = Rank.NOVICE,
        val unlockedThemes: Set<String> = setOf("default"),
        val unlockedMusic: Set<String> = setOf("theme_a"),
        val unlockedPieceStyles: Set<String> = setOf("standard"),
        val unlockedEffects: Set<String> = setOf("basic"),
        val statistics: PlayerStatistics = PlayerStatistics()
    )
    
    enum class Rank(val minLevel: Int, val displayName: String, val color: String) {
        NOVICE(1, "Novice", "#808080"),
        APPRENTICE(10, "Apprentice", "#CD7F32"),
        ADEPT(20, "Adept", "#C0C0C0"),
        EXPERT(30, "Expert", "#FFD700"),
        MASTER(40, "Master", "#E5E4E2"),
        GRANDMASTER(50, "Grandmaster", "#50C878"),
        LEGEND(60, "Legend", "#FF4500"),
        MYTHIC(70, "Mythic", "#9966CC"),
        DIVINE(85, "Divine", "#00FFFF"),
        ETERNAL(100, "Eternal", "#FFD700")
    }
    
    data class PlayerStatistics(
        val totalGamesPlayed: Int = 0,
        val totalLinesCleared: Int = 0,
        val totalScore: Long = 0,
        val highestScore: Int = 0,
        val totalTSpins: Int = 0,
        val totalTetrises: Int = 0,
        val totalPerfectClears: Int = 0,
        val longestCombo: Int = 0,
        val totalPlayTime: Long = 0,
        val favoriteMode: String = "Classic",
        val winRate: Float = 0f
    )
    
    data class Unlockable(
        val id: String,
        val name: String,
        val type: UnlockableType,
        val requiredLevel: Int,
        val description: String,
        val previewUrl: String? = null
    )
    
    enum class UnlockableType {
        THEME, MUSIC, PIECE_STYLE, EFFECT
    }
    
    private val _progressionState = MutableStateFlow(ProgressionState())
    val progressionState: StateFlow<ProgressionState> = _progressionState.asStateFlow()
    
    private val unlockables = listOf(
        // Themes
        Unlockable("cyberpunk", "Cyberpunk", UnlockableType.THEME, 1, "Neon lights and digital rain"),
        Unlockable("retro", "Retro 8-bit", UnlockableType.THEME, 5, "Classic arcade nostalgia"),
        Unlockable("nature", "Nature", UnlockableType.THEME, 10, "Calming natural colors"),
        Unlockable("minimal", "Minimal", UnlockableType.THEME, 15, "Clean and simple design"),
        Unlockable("galaxy", "Galaxy", UnlockableType.THEME, 25, "Cosmic space theme"),
        Unlockable("matrix", "Matrix", UnlockableType.THEME, 35, "Digital rain effect"),
        Unlockable("rainbow", "Rainbow", UnlockableType.THEME, 50, "Vibrant rainbow colors"),
        
        // Music
        Unlockable("chiptune", "Chiptune", UnlockableType.MUSIC, 3, "8-bit retro music"),
        Unlockable("synthwave", "Synthwave", UnlockableType.MUSIC, 8, "80s electronic vibes"),
        Unlockable("orchestral", "Orchestral", UnlockableType.MUSIC, 20, "Epic classical score"),
        Unlockable("jazz", "Jazz", UnlockableType.MUSIC, 30, "Smooth jazz rhythms"),
        Unlockable("metal", "Metal", UnlockableType.MUSIC, 40, "Heavy metal intensity"),
        Unlockable("lofi", "Lo-fi", UnlockableType.MUSIC, 60, "Relaxing lo-fi beats"),
        
        // Piece Styles
        Unlockable("glass", "Glass", UnlockableType.PIECE_STYLE, 7, "Transparent glass pieces"),
        Unlockable("pixel", "Pixel", UnlockableType.PIECE_STYLE, 12, "Chunky pixel art"),
        Unlockable("hologram", "Hologram", UnlockableType.PIECE_STYLE, 22, "Futuristic holograms"),
        Unlockable("crystal", "Crystal", UnlockableType.PIECE_STYLE, 32, "Shimmering crystals"),
        Unlockable("animated", "Animated", UnlockableType.PIECE_STYLE, 45, "Living pieces"),
        
        // Effects
        Unlockable("particles", "Particles", UnlockableType.EFFECT, 5, "Particle explosions"),
        Unlockable("trails", "Trails", UnlockableType.EFFECT, 10, "Motion trails"),
        Unlockable("explosions", "Explosions", UnlockableType.EFFECT, 18, "Explosive line clears"),
        Unlockable("lightning", "Lightning", UnlockableType.EFFECT, 28, "Electric effects"),
        Unlockable("shatter", "Shatter", UnlockableType.EFFECT, 38, "Shattering animations"),
        Unlockable("quantum", "Quantum", UnlockableType.EFFECT, 55, "Quantum particle effects")
    )
    
    fun addXP(amount: Int, source: String = "gameplay") {
        val currentState = _progressionState.value
        var newXP = currentState.currentXP + amount
        var newTotalXP = currentState.totalXP + amount
        var newLevel = currentState.level
        var leveledUp = false
        
        // Check for level up
        while (newXP >= currentState.xpToNextLevel && newLevel < MAX_LEVEL) {
            newXP -= currentState.xpToNextLevel
            newLevel++
            leveledUp = true
            
            Timber.d("Level up! New level: $newLevel")
            checkUnlocks(newLevel)
        }
        
        // Calculate XP needed for next level
        val xpToNext = calculateXPForLevel(newLevel)
        
        // Update rank
        val newRank = calculateRank(newLevel)
        
        _progressionState.value = currentState.copy(
            level = newLevel,
            currentXP = newXP,
            totalXP = newTotalXP,
            xpToNextLevel = xpToNext,
            rank = newRank
        )
        
        if (leveledUp) {
            onLevelUp(newLevel)
        }
        
        Timber.d("Added $amount XP from $source. Total: $newTotalXP")
    }
    
    private fun calculateXPForLevel(level: Int): Int {
        return (XP_PER_LEVEL_BASE * XP_PER_LEVEL_MULTIPLIER.toDouble().pow((level - 1).toDouble())).toInt()
    }
    
    private fun calculateRank(level: Int): Rank {
        return Rank.values()
            .sortedByDescending { it.minLevel }
            .first { level >= it.minLevel }
    }
    
    private fun checkUnlocks(level: Int) {
        val currentState = _progressionState.value
        val newUnlocks = unlockables.filter { 
            it.requiredLevel == level 
        }
        
        val themes = currentState.unlockedThemes.toMutableSet()
        val music = currentState.unlockedMusic.toMutableSet()
        val pieceStyles = currentState.unlockedPieceStyles.toMutableSet()
        val effects = currentState.unlockedEffects.toMutableSet()
        
        newUnlocks.forEach { unlock ->
            when (unlock.type) {
                UnlockableType.THEME -> themes.add(unlock.id)
                UnlockableType.MUSIC -> music.add(unlock.id)
                UnlockableType.PIECE_STYLE -> pieceStyles.add(unlock.id)
                UnlockableType.EFFECT -> effects.add(unlock.id)
            }
            
            Timber.d("Unlocked: ${unlock.name} (${unlock.type})")
        }
        
        _progressionState.value = currentState.copy(
            unlockedThemes = themes,
            unlockedMusic = music,
            unlockedPieceStyles = pieceStyles,
            unlockedEffects = effects
        )
    }
    
    private fun onLevelUp(newLevel: Int) {
        // Trigger level up effects and notifications
        when (newLevel) {
            10, 20, 30, 40, 50, 60, 70, 85, 100 -> {
                // Major milestone reached
                Timber.d("Major milestone reached: Level $newLevel")
            }
        }
    }
    
    fun updateStatistics(stats: PlayerStatistics) {
        _progressionState.value = _progressionState.value.copy(statistics = stats)
    }
    
    fun processGameResults(
        score: Int,
        lines: Int,
        level: Int,
        gameMode: String,
        tSpins: Int = 0,
        tetrises: Int = 0,
        perfectClears: Int = 0,
        maxCombo: Int = 0,
        duration: Long = 0
    ) {
        // Calculate XP earned
        var xpEarned = 0
        xpEarned += lines * XP_PER_LINE
        xpEarned += tetrises * XP_PER_TETRIS
        xpEarned += tSpins * XP_PER_T_SPIN
        xpEarned += perfectClears * XP_PER_PERFECT_CLEAR
        xpEarned += maxCombo * XP_PER_COMBO
        xpEarned += (level - 1) * XP_PER_LEVEL_UP
        
        // Bonus XP based on score
        xpEarned += score / 100
        
        addXP(xpEarned, "game_complete")
        
        // Update statistics
        val currentStats = _progressionState.value.statistics
        val newStats = currentStats.copy(
            totalGamesPlayed = currentStats.totalGamesPlayed + 1,
            totalLinesCleared = currentStats.totalLinesCleared + lines,
            totalScore = currentStats.totalScore + score,
            highestScore = maxOf(currentStats.highestScore, score),
            totalTSpins = currentStats.totalTSpins + tSpins,
            totalTetrises = currentStats.totalTetrises + tetrises,
            totalPerfectClears = currentStats.totalPerfectClears + perfectClears,
            longestCombo = maxOf(currentStats.longestCombo, maxCombo),
            totalPlayTime = currentStats.totalPlayTime + duration
        )
        updateStatistics(newStats)
    }
    
    fun getUnlockedItems(): Map<UnlockableType, List<Unlockable>> {
        val currentState = _progressionState.value
        return unlockables
            .filter { unlock ->
                when (unlock.type) {
                    UnlockableType.THEME -> unlock.id in currentState.unlockedThemes
                    UnlockableType.MUSIC -> unlock.id in currentState.unlockedMusic
                    UnlockableType.PIECE_STYLE -> unlock.id in currentState.unlockedPieceStyles
                    UnlockableType.EFFECT -> unlock.id in currentState.unlockedEffects
                }
            }
            .groupBy { it.type }
    }
    
    fun getAvailableUnlocks(currentLevel: Int): List<Unlockable> {
        return unlockables.filter { it.requiredLevel <= currentLevel }
    }
    
    fun getNextUnlocks(): List<Unlockable> {
        val currentLevel = _progressionState.value.level
        return unlockables
            .filter { it.requiredLevel > currentLevel }
            .sortedBy { it.requiredLevel }
            .take(3)
    }
    
    fun reset() {
        _progressionState.value = ProgressionState()
    }
}